### DISCUSSION TITLE
How is contribution possible (why custom branch is not installable)?

### DISCUSSION THREAD

**Discussion by programmador** [stephenh/ts-proto#977] [Q&A]
Reproduction:
```
~ $ mkdir temp
~ $ cd temp
~/temp $ yarn init
yarn init v1.22.19
question name (temp):
question version (1.0.0):
question description:
question entry point (index.js):
question repository url:
question author:
question license (MIT):
question private:
success Saved package.json
Done in 8.59s.
~/temp $ yarn add ts-proto@git+ssh://git@github.com:stephenh/ts-proto.git#v1.151.1
```

Gives an error:
```
Internal Error: ts-proto@workspace:.: This package doesn't seem to be present in your lockfile; run "yarn install" to update the lockfile
```

While just `yarn add ts-proto@1.151.1` successfully installs the package.

I've tried to install the lib from a folder `yarn add ts-proto/path-to-folder` but this way codegen does not work as yarn skips build phase for some reason. It just copies the lib in node_modules without trying to build it. Though it seems like when I build the lib beforehand - yarn also copies the `build` folder successfully.

BTW while installing via full git+ssh link I also get a warning: "Workspaces can only be enabled in private projects."

**Comment by stephenh** [comment]
Hi @programmador ; I've not actually used any of the `yarn add ...git...` or `yarn add ...folder...` approaches; back in the day, I'd tried to use `yarn link` / `npm link` for this, but they never worked 100%.

So instead I've always just done it by hand, like:

* Run `yarn build -w` in ts-proto in one terminal
* In my real project run `yarn watch-ts-proto` in another terminal that does something like:

```
    "watch-ts-proto": "copy-and-watch --watch ../ts-proto/build/ ./node_modules/ts-proto/build/"
```

If you figure out how to get the git- or path-based approaches for yarn working with ts-proto, that'd be great!

**Reply by programmador** [reply]
path-based approch is actually working. You just have to run `yarn build` in ts-proto manually.
After that You can `yarn add ts-proto@./path/to/ts-proto`
Though probably it's not the most convenient way as it doesn't `watch`.

**Comment by programmador** [comment]
Maybe the reason of unability to install custom branch is related to this warning:
```
Internal Error: ts-proto@workspace:.: This package doesn't seem to be present in your lockfile; run "yarn install" to update the lockfile
```
It seems like some of the dependencies has `workspaces` section somehow broken so that installation into a project without workspaces is not available. BTW I didn't try to setup workspaces somehow.

**Comment by stephenh** [comment]
I've added a note about `yarn add` to the readme; thanks!
